---
feature_text: |
  ## Genomic Visualization and Interpretations
title: Variant annotation with VEP
categories:
    - Module 5
feature_image: "assets/genvis-dna-bg_optimized_v1a.png"
date: 0005-02-01
---

Often it will be informative to annotate variants with additional information in order to get a sense of a variants impact on a phenotype. One tool that makes this process quick and straightforward is the ensembl [Variant Effect Predictor](http://www.ensembl.org/info/docs/tools/vep/index.html) (VEP). This program is available both as a stand alone software program based in [perl](https://www.perl.org/) and as a web based GUI. In this module we will learn how to use [VEP](http://www.ensembl.org/info/docs/tools/vep/index.html) in it's raw perl form.

### Installing perl
##### MAC and UNIX
In order to use [VEP](http://www.ensembl.org/info/docs/tools/vep/index.html) we will first need to download and install [perl](https://www.perl.org/), a high level scripting language. First let's check if you have perl â‰¥ 5.10 already installed, open a command prompt "terminal" on your local machine and run the code below.
```bash
# check perl version
perl -v
```
If you see a message to the effect of "This is perl 5" then you can ignore the next bit, otherwise you will need to download and install perl. To do this navigate to the perl downloads page at [https://www.perl.org/get.html](https://www.perl.org/get.html) and select "get started" for your specific operating system.

{% include figure.html image="/assets/VEP/downloadPerl.png" width="950" %}

Then select activeState perl, once the installer is downloaded follow the on screen instructions and check your install with `perl -v`.

{% include figure.html image="/assets/VEP/DownloadActivestatePerl.png" width="950" %}

##### Windows
In order to use [VEP](http://www.ensembl.org/info/docs/tools/vep/index.html) on widnows we will first need to download and install a special flavor of perl called DWIMperl. Navigate to [http://dwimperl.com/windows.html](http://dwimperl.com/windows.html) and download the "Dwimperl-5.14.2.1-v7-32.exe" executable at the bottom of the page. Then run the executable and follow the on screen instructions. Once finished search and open **Command Prompt** from the "Start Menu" and type `perl -v` to check that the installation was successful.

{% include figure.html image="/assets/VEP/downloadDWIMperl.png" width="950" %}

### Installing VEP
With [perl](https://www.perl.org/) now installed we can go ahead and download and install [VEP](http://www.ensembl.org/info/docs/tools/vep/index.html) itself. The recommended way to do this is by cloning the github repo available [here](https://github.com/Ensembl/ensembl-vep). We won't actually be cloning the repo, though you certainly could. Instead select the green **Clone or download** button to the right and then select **Download Zip**. Once the file is downloaded go ahead and unzip it.

{% include figure.html image="/assets/VEP/downloadVep.png" width="750" %}

Once that is complete you will need to navigate to the directory where you unziped the [VEP](http://www.ensembl.org/info/docs/tools/vep/index.html) repo and run the code below.

```bash
# change to directory where vep is
cd ./ensembl-vep-release-90

# initate VEP installations
perl INSTALL.pl --NO_HTSLIB --NO_TEST
```

This will start the [VEP](http://www.ensembl.org/info/docs/tools/vep/index.html) installation process. You will be asked if you want to "download local cache files", "download fasta files", and "install plugins". Because of their size we will answer "no" to all these however you can always change this by re-running `perl INSTALL.pl`.

{% include figure.html image="/assets/VEP/vepInstall.png" width="750" %}

Finally let's check our [VEP](http://www.ensembl.org/info/docs/tools/vep/index.html) installation, from the same directory you ran `perl INSTALL.pl` run `./vep --help`. If everything went okay you should see some [VEP](http://www.ensembl.org/info/docs/tools/vep/index.html) version numbers appear.

{% include figure.html image="/assets/VEP/vepVersion.png" width="750" %}

### Running VEP
As we have mentioned previously [VEP](http://www.ensembl.org/info/docs/tools/vep/index.html) can either be run via the command line or through a web GUI. For the remainder of this section we will be doing both side by side however we should note that using [VEP](http://www.ensembl.org/info/docs/tools/vep/index.html) from the command line is more flexible and there is a greater range of features.

To start let's go ahead and annotate the germline variant [rs1799966](https://www.ncbi.nlm.nih.gov/SNP/snp_ref.cgi?type=rs&rs=rs1799966) using the default [ensembl input format](https://www.ensembl.org/info/docs/tools/vep/vep_formats.html). Essentially we need to create a file with 5 columns corresponding to "chromosome", "start" (1-based), "stop" (1-based), "reference/variant", and "strand" (corresponding to reference/variant). You can optionally have a sixth column to add a unique identifier for this row. After creating the file `variant.txt` we run [VEP](http://www.ensembl.org/info/docs/tools/vep/index.html) with the following options.
1. `-i` input file, the file format is automatically detected
2. `-o` output file to write results
3. `--database` make queries to public ensembl databases instead of looking for local copies
4. `--species` species for which annotations should be obtained
5. `--everything` flag to output additional annotations

```R
# make a file with a single variant in
echo "17 43071077 43071077 A/G - variant_1" > variant.txt

# run VEP
./vep -i variant.txt -o variant.anno.txt --database --species "human" --everything
```

We can do the same thing by inputing this variant via the the web interface and clicking on **Run**.

{% include figure.html image="/assets/VEP/runVep_v1.png" width="750" %}

This will submit the job to ensembl servers, the page will refresh every few seconds. When the job completes click on **view results**.

{% include figure.html image="/assets/VEP/runVep_v2.png" width="750" %}

Doing so will take you to a web page where you will be able to view summary statistics, the results, and options to filter or export a file.

{% include figure.html image="/assets/VEP/runVep_v3.png" width="750" %}

You have probably noticed from your exploration of the results that even though only one variant was supplied multiple rows were output. By default [VEP](http://www.ensembl.org/info/docs/tools/vep/index.html) returns annotations for each "transcript" and each "variant consequence". While this is informative it is often desireable to only have one annotation per gene. We can achieve this on our command line by adding the `--per_gene` parameter. The same thing can be achieved through the web interface by expanding the **filtering options** tab and setting **Restrict results** to **Show one selected consequence per gene**. Go ahead and do that now through either the command line or web GUI.

{% include figure.html image="/assets/VEP/runVep_v4.png" width="750" %}

### Exercises

Now that you have an inital [VEP](http://www.ensembl.org/info/docs/tools/vep/index.html) result try and answer a few questions about your data. If you need help in understanding a certain column hover over it with your cursor or look at the [VEP](http://www.ensembl.org/info/docs/tools/vep/index.html) documentation available [here](http://www.ensembl.org/info/docs/tools/vep/vep_formats.html#output).

What is the maximum allele frequency observed for this vairant in the 1000 genomes european population?
{% include question.html question="Answer" answer='0.3598'%}

In what gene is this variant located?
{% include question.html question="Answer" answer='BRCA1'%}

What is the amino acid change for this variant?
{% include question.html question="Answer" answer='S/G'%}
